src/library/scala/collection/Iterator.scala

```diff
@@ -1115,55 +1115,59 @@ object Iterator extends IterableFactory[Iterator] {
   }
 
   /** Creates an iterator to which other iterators can be appended efficiently.
-    *  Nested ConcatIterators are merged to avoid blowing the stack.
-    */
+   *  Nested ConcatIterators are merged to avoid blowing the stack.
+   */
   private final class ConcatIterator[+A](private var current: Iterator[A @uncheckedVariance]) extends AbstractIterator[A] {
     private var tail: ConcatIteratorCell[A @uncheckedVariance] = null
     private var last: ConcatIteratorCell[A @uncheckedVariance] = null
     private var currentHasNextChecked = false
 
-    // Advance current to the next non-empty iterator
-    // current is set to null when all iterators are exhausted
-    @tailrec
-    private[this] def advance(): Boolean = {
-      if (tail eq null) {
-        current = null
-        last = null
-        false
-      }
-      else {
-        current = tail.headIterator
-        tail = tail.tail
-        merge()
-        if (currentHasNextChecked) true
-        else if ((current ne null) && current.hasNext) {
-          currentHasNextChecked = true
-          true
-        } else advance()
+    @`inline` private def mergeAndAdvance(): Boolean = {
+      // Advance current to the next non-empty iterator
+      // current is set to null when all iterators are exhausted
+      @tailrec def advance(): Boolean = {
+        if (tail == null) {
```

Comment:
Is there a clear criterion to choose `eq null` vs `== null`? Throughout the code-base, the `eq` is used to compare objects by reference equality, but would it also be more efficient here?

Reply:
`x == null` is always compiled down to the exact same bytecode (or .sjsir, or .nir) as `x eq null`. There is no efficiency argument.

Reply:
I was motivated by @SethTisue previous comment on a previous PR to this code, in which he requested `==` as a matter of style.

Reply:
at https://github.com/scala/scala/pull/7219

Reply:
Ok. 

I am asking because this PR has a bit of whitespace and diff noise. 

```Scala

1136:           if (last eq tail) last = last.tail

1154:             if (last == null) last = c.last
```

Reply:
There are two commits to separate wheat and chaff.